### PR TITLE
Dns response 2198 v1

### PR DESCRIPTION
**Pull request by StianHB**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues/2198) ticket:2198

Describe changes:
- Created the dns_response keyword that allows matching in rules in A record responses in DNS
- Updated the user-docs to reflect the new keyword

### CODE REVIEW

**Comment by victorjulien** [comment]
Various issues that will need to be addressed:
- QA failing
- not in sync with master ,please rebase to current master
- file that has no place in our tree src/CMakeLists.txt
- lots of debug code still in place

File: src/app-layer-dns-common.c [hunk 1]
```diff
@@ -851,10 +851,10 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
...
         {
             if (datalen == 0 || datalen == 4) {
-                //PrintRawDataFp(stdout, data, ntohs(head->len));
-                //char a[16];
-                //PrintInet(AF_INET, (const void *)data, a, sizeof(a));
-                //SCLogInfo("A %s TTL %u", a, ntohl(head->ttl));
+               // PrintRawDataFp(stdout, data, ntohs(head->len));
+                char a[16];
+                PrintInet(AF_INET, (const void *)data, a, sizeof(a));
+                SCLogInfo("A %s TTL %u", a, ntohl(head->ttl));
```
Comment (victorjulien) [inline src/app-layer-dns-common.c]: debug code

File: src/detect-dns-query.c [hunk 2]
```diff
@@ -92,14 +92,9 @@ void DetectDnsQueryRegister (void)
     DetectAppLayerInspectEngineRegister("dns_request",
             ALPROTO_DNS, SIG_FLAG_TOSERVER, 1,
             DetectEngineInspectDnsRequest);
-    DetectAppLayerInspectEngineRegister("dns_response",
```
Comment (victorjulien) [inline src/detect-dns-query.c]: removing this will break other things I fear, thinking mostly about lua scripts

File: src/detect-dns-response.c [hunk 3]
```diff
@@ -0,0 +1,212 @@
+/* Copyright (C) 2013 Open Information Security Foundation
```
Comment (victorjulien) [inline src/detect-dns-response.c]: update year

File: src/detect-engine-dns.c [hunk 4]
```diff
@@ -119,6 +120,48 @@ int DetectEngineInspectDnsQueryName(ThreadVars *tv,
...
+                                    sizeof(DNSAnswerEntry) + answer->fqdn_len);
+
+            char a[16] = "";
+            PrintInet(AF_INET, (const void *) temp, a, sizeof(a));
+
+            uint8_t *ipstring = (uint8_t *) a;
+
+            r = DetectEngineContentInspection(de_ctx, det_ctx,
+                                              s, smd,
+                                              f, ipstring, sizeof(a), 0,
```
Comment (victorjulien) [inline src/detect-engine-dns.c]: seems you should pass 'strlen(a)' here, as the a is a string and we should pass the length of the actual data. Now you pass 16 even if we have '1.1.1.1' == 7

File: src/detect-engine-dns.c [hunk 5]
```diff
@@ -169,6 +212,35 @@ static void PrefilterTxDnsQuery(DetectEngineThreadCtx *det_ctx,
...
+            const uint8_t *temp = (const uint8_t *) ((uint8_t *) answer + sizeof(DNSAnswerEntry) + answer->fqdn_len);
+
+            char a[16] = "";
+            PrintInet(AF_INET, (const void *) temp, a, sizeof(a));
+
+            uint8_t *ipstring = (uint8_t *)a;
+
+            if (sizeof(a) >= mpm_ctx->minlen) {
+                (void) mpm_table[mpm_ctx->mpm_type].Search(mpm_ctx, &det_ctx->mtcu, &det_ctx->pmq,
+                                                           ipstring, sizeof(a));
```
Comment (victorjulien) [inline src/detect-engine-dns.c]: same here

File: src/detect-dns-query.c (hunk 2 above)
Comment (StianHB) [reply src/detect-dns-query.c]: It is just moved to detect-dns-response.c, I think that is more correct then leaving it in detect-dns-query.c, do you want me to move it back to detect-dns-query.c ?

**Comment by victorjulien** [comment]
Please close a PR when you replace it, thanks!

Replaced by #3006

**Pull request closed without merging**